File: backend/app/models/calendar.py
import uuid
from datetime import datetime, timezone
from sqlalchemy import String, Boolean, DateTime, ForeignKey, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.core.database import Base


class Calendar(Base):
    __tablename__ = "calendars"

    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    user_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    color: Mapped[str] = mapped_column(String(7), default="#3B82F6")  # hex color
    is_default: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )

    # Relationships
    owner: Mapped["User"] = relationship("User", back_populates="calendars")  # noqa: F821
    events: Mapped[list["Event"]] = relationship(
        "Event", back_populates="calendar", cascade="all, delete-orphan"
    )


class Event(Base):
    __tablename__ = "events"

    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    calendar_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("calendars.id", ondelete="CASCADE"), nullable=False, index=True
    )
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    start_time: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    end_time: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    recurrence_rule: Mapped[str | None] = mapped_column(String(500), nullable=True)
    location: Mapped[str | None] = mapped_column(String(300), nullable=True)
    color: Mapped[str | None] = mapped_column(String(7), nullable=True)
    is_all_day: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

    # Relationships
    calendar: Mapped["Calendar"] = relationship("Calendar", back_populates="events")
    tasks: Mapped[list["Task"]] = relationship(  # noqa: F821
        "Task", back_populates="linked_event"
    )

File: mobile/lib/shared/models/calendar.dart
import 'package:flutter/material.dart';

class CalendarModel {
  const CalendarModel({
    required this.id,
    required this.userId,
    required this.name,
    required this.colorHex,
    required this.isDefault,
    required this.createdAt,
  });

  final String id;
  final String userId;
  final String name;
  final String colorHex;
  final bool isDefault;
  final DateTime createdAt;

  Color get color {
    final hex = colorHex.replaceFirst('#', '');
    return Color(int.parse('FF$hex', radix: 16));
  }

  factory CalendarModel.fromJson(Map<String, dynamic> json) => CalendarModel(
        id: json['id'] as String,
        userId: json['user_id'] as String,
        name: json['name'] as String,
        colorHex: json['color'] as String,
        isDefault: json['is_default'] as bool,
        createdAt: DateTime.parse(json['created_at'] as String),
      );
}

class EventModel {
  const EventModel({
    required this.id,
    required this.calendarId,
    required this.title,
    required this.startTime,
    required this.endTime,
    this.description,
    this.location,
    this.colorHex,
    this.isAllDay = false,
    this.recurrenceRule,
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String calendarId;
  final String title;
  final DateTime startTime;
  final DateTime endTime;
  final String? description;
  final String? location;
  final String? colorHex;
  final bool isAllDay;
  final String? recurrenceRule;
  final DateTime createdAt;
  final DateTime updatedAt;

  Color? get color {
    if (colorHex == null) return null;
    final hex = colorHex!.replaceFirst('#', '');
    return Color(int.parse('FF$hex', radix: 16));
  }

  factory EventModel.fromJson(Map<String, dynamic> json) => EventModel(
        id: json['id'] as String,
        calendarId: json['calendar_id'] as String,
        title: json['title'] as String,
        startTime: DateTime.parse(json['start_time'] as String).toLocal(),
        endTime: DateTime.parse(json['end_time'] as String).toLocal(),
        description: json['description'] as String?,
        location: json['location'] as String?,
        colorHex: json['color'] as String?,
        isAllDay: json['is_all_day'] as bool? ?? false,
        recurrenceRule: json['recurrence_rule'] as String?,
        createdAt: DateTime.parse(json['created_at'] as String),
        updatedAt: DateTime.parse(json['updated_at'] as String),
      );
}
